### PR TITLE
Move TorchDynamo into PyTorch core

### PR DESCRIPTION
**Pull request by jansel**
Context:
https://github.com/pytorch/torchdynamo/issues/1588

This PR moves [TorchDynamo](https://github.com/pytorch/torchdynamo) and TorchInductor into PyTorch core.
- `torchdynamo` becomes `torch._dynamo`
- `torchinductor` becomes `torch._inductor`

This PR was generated by running `copy_to_core.sh` in https://github.com/pytorch/torchdynamo/pull/1538

### CODE REVIEW

File: torch/csrc/dynamo/guards.cpp [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+    "torch._C._dynamo.guards",
+    "Module containing checks on tensors",
+    -1,
+    _methods};
+
+} // namespace
+
+PyObject* torch_c_dynamo_guards_init() {
+  // initialize TensorGuardsType
+  TensorGuardsType.tp_name = "torchdynamo._guards.TensorGuards";
```
Comment (albanD) [inline torch/csrc/dynamo/guards.cpp:352]: Are we going to keep the `torchdynamo` top level module? Or should this be in `torch._dynamo`.

File: torch/csrc/Module.cpp [hunk 2]
```diff
@@ -1132,6 +1133,7 @@ PyObject* initModule() {
   torch::jit::initJITBindings(module);
   torch::monitor::initMonitorBindings(module);
   torch::impl::dispatch::initDispatchBindings(module);
+  torch::dynamo::initDynamoBindings(module);
```
Comment (albanD) [inline torch/csrc/Module.cpp:1136]: Isn't this going to conflict with the dynamo that is provided by the `torchdynamo` package?

File: test/dynamo/test_aot_autograd.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
+# Owner(s): ["module: dynamo"]
```
Comment (albanD) [inline test/dynamo/test_aot_autograd.py:1]: Are all these test running? I don't see any exclude for these?

File: torch/csrc/Module.cpp (hunk 2 above)
Comment (jansel) [reply torch/csrc/Module.cpp:1136]: The torchdynamo package is going away.

File: test/dynamo/test_aot_autograd.py (hunk 3 above)
Comment (jansel) [reply test/dynamo/test_aot_autograd.py:1]: Yes, currently they are running.  I think it makes sense to disable them before landing though, and turn them on in another PR.

File: torch/csrc/dynamo/guards.cpp (hunk 1 above)
Comment (jansel) [reply torch/csrc/dynamo/guards.cpp:352]: Good catch, will fix. This is actually `torch._C._dynamo.guards.TensorGuards` now.

File: torch/csrc/Module.cpp (hunk 2 above)
Comment (albanD) [reply torch/csrc/Module.cpp:1136]: What about the transition time during which the version in core is "disabled" and `torchdynamo` remains the source of truth?

Comment (jansel) [reply torch/csrc/Module.cpp:1136]: https://github.com/pytorch/torchdynamo/pull/1538 makes top level torchdynamo use `torch._C._dynamo` for the C++ bits.

**Review by voznesenskym — APPROVED** [review]

**Pull request closed without merging**